### PR TITLE
try to keep memory region alive untill julia handled it.

### PR DESCRIPTION
**Pull request by vchuravy**
@piiswrong @pluskid I started addressing native operators in Julia again and I have run into the issue that the memory areas passed to Julia for tags and arrays went out of scope and contained garbage, before Julia got to them. 

This PR is my attempt at fixing this, but it still doesn't work. What would be the best way of waiting till the async operation has finished to then `delete` the pointers?

### CODE REVIEW

**Comment by tqchen** [comment]
The best way is to push a write operation to the engine, write to the variable of the array, with the lambda function being delete

**Comment by tqchen** [comment]
see https://github.com/dmlc/mxnet/blob/master/include/mxnet/ndarray.h#L330
on how ndarray delete its memory, the lambda function will only be executed after the dependency is fullfilled

**Comment by vchuravy** [comment]
@tqchen The problem is not with the NDArrays (yet), but with the `std::vectors` that store the pointers and tags to the NDArrays.

File: src/operator/ndarray_op.cc [hunk 1]
```diff
@@ -80,40 +83,43 @@ void NDArrayOp<xpu>::Backward(const OpContext &ctx,
...
+  for (auto& i : *ptrs) {
     ndcpy.push_back(*reinterpret_cast<NDArray*>(i));
   }
 
-  CHECK(param_.pinfo->backward(ptrs.size(), ptrs.data(), tags.data(), param_.pinfo->p_backward));
-  Engine::Get()->PushSync([ndcpy, ctx](RunContext rctx){ ctx.async_on_complete(); },
-                          ndctx, ndvar, {});
+  CHECK(param_.pinfo->backward(ptrs->size(), ptrs->data(), tags->data(), param_.pinfo->p_backward));
+  Engine::Get()->PushSync([ptrs, tags, ndcpy, ctx](RunContext rctx){
+      ctx.async_on_complete();
```
Comment (tqchen) [inline src/operator/ndarray_op.cc]: when you call PushSync, there is no need to call async_on_compelte, this is only reserved for the async call in the operators

Comment (tqchen) [inline src/operator/ndarray_op.cc]: Oh my bad, please ignore previous comment

**Comment by tqchen** [comment]
So here is the problem, because call into julia is async (aha), the PushSync on julia may not necessarily  happen before the last PushSync which notifies the completion and finishes all the jobs.  

This is a really tricky thing and I guess the only solution maybe offer another callback to Julia to enable it notify its completion in the async call.

**Comment by tqchen** [comment]
@vchuravy Do you have any updates on this issue?

**Comment by tqchen** [comment]
Closing this now, but feel free to re-open. I think a callback from Julia side is necessary for this task.

**Comment by tqchen** [comment]
OK, bring in @piiswrong 

I think there is one ways to resolve this issue: modify the CustomOp API so that it contains a callback to notify the finish of the operation. I realized that this is necessary for asynchronize operations.

Specifically, the ```Forward```, and ```Backward``` function will take an argument that is a C style callback API, and the user is able to call the callback when the Forward and Backward operations finishes on the native side

**Comment by vchuravy** [comment]
@tqchen I hope to get to this eventually, but I am planning to wait on Julia's multithreading support.

The solution you outlined sounds feasible. The memory management part is quite tricky here.

**Pull request closed without merging**